gridscale/resource_gridscale_bucket.go

```
@@ -63,15 +66,108 @@ func resourceGridscaleBucket() *schema.Resource {
+							Required: true,
+						},
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"prefix": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"expiration_days": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  365,
+						},
+						"noncurrent_version_expiration_days": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  365,
+						},
+						"incomplete_upload_expiration_days": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  3,
+						},
+					},
+				},
+			},
 		},
 		Timeouts: &schema.ResourceTimeout{
 			Create: schema.DefaultTimeout(5 * time.Minute),
 			Delete: schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(5 * time.Minute),
 		},
 	}
 }
 
 func resourceGridscaleBucketRead(d *schema.ResourceData, meta interface{}) error {
+	s3Host := d.Get("s3_host").(string)
+	accessKey := d.Get("access_key").(string)
+	secretKey := d.Get("secret_key").(string)
+	bucketName := d.Get("bucket_name").(string)
+
+	s3Client := initS3Client(&gridscaleS3Provider{
+		AccessKey: accessKey,
+		SecretKey: secretKey,
+	}, s3Host)
+
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout(schema.TimeoutRead))
+	defer cancel()
+
+	// Fetch lifecycle configuration
+	output, err := s3Client.GetBucketLifecycleConfigurationWithContext(ctx, &s3.GetBucketLifecycleConfigurationInput{
+		Bucket: aws.String(bucketName),
+	})
+	if err != nil {
+		var aerr awserr.Error
+		if errors.As(err, &aerr) && aerr.Code() == "NoSuchLifecycleConfiguration" {
+			// If the error indicates no lifecycle configuration exists, set the lifecycle_rule attribute to nil
+			d.Set("lifecycle_rule", nil)
+		} else {
+			// For any other error, return a formatted error message with context
+			return fmt.Errorf("error reading lifecycle configuration for bucket %s: %w", bucketName, err)
+		}
```

~I think here we can simply return the error. 2 points:~
- ~failed `Read` will not cause rollback (so it is safe).~
- ~`d.Set("lifecycle_rule", nil)` will cause inconsistency between the state and the tf config.~